### PR TITLE
Add API docs for ONNX Runtime and Intel

### PR DESCRIPTION
**Pull request by lewtun**
# What does this PR do?

This PR adds the API docs for the `intel` and `onnxruntime` packages. It depends on the fixes to `doc-builder` in [this PR](https://github.com/huggingface/doc-builder/pull/149), so the CI should pass once that's merged.

The docs are structured as follows:

```
docs
└── source
    ├── index.mdx  # The landing page / README
    ├── intel  # API docs go under their own subpackage
    │   ├── configuration.mdx  # I use the same naming conventions across subpackages
    │   ├── optimization.mdx
    │   ├── pruning.mdx
    │   ├── quantization.mdx
    │   └── trainer.mdx
    ├── onnxruntime
    │   ├── configuration.mdx
    │   ├── optimization.mdx
    │   ├── quantization.mdx
    │   └── trainer.mdx
    ├── quickstart.mdx  # Guides go in the root of source/
    └── _toctree.yml  # The table of contents
```

While putting together the docs, I ran into a few `optimum` quirks that I would like feedback on:

1. Unlike `transformers`, `optimum` doesn't have a dev install that includes the `intel` and `onnxruntime` packages. Is this by design? Would it make sense to have `pip install '.[dev]'` install all subpackages?
2. Some of the modules have inconsistent naming across `intel` and `onnxruntime`. For example, we have `intel.neural_compressor.optimizer` and `onnxruntime.optimization`. I've standardised this in the docs to have an "optimization" section per subpackage, and am happy to rename the `intel` module name if you want.
3. To build the docs, I had to include the `intel.neural_compressor` imports in the root init of the `intel` subpackage. Is this OK?

### CODE REVIEW

File: .github/workflows/build_dev_documentation.yml [hunk 1]
```diff
@@ -48,7 +48,7 @@ jobs:
           cd ..
 
           cd optimum
-          pip install .[dev]
+          pip install .[dev,onnxruntime,intel]
```
Comment (lewtun) [inline .github/workflows/build_dev_documentation.yml:51]: I have to include all the dependencies here to build the docs

File: optimum/intel/__init__.py [hunk 2]
```diff
@@ -11,3 +11,9 @@
 #  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
+
+from .neural_compressor.config import IncConfig, IncPruningConfig, IncQuantizationConfig
```
Comment (lewtun) [inline optimum/intel/__init__.py:15]: Here I have to include the `neural_compressor` imports in the root init of the `intel` subpackage to build the docs

File: docs/source/intel/optimization.mdx [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+-->
+
+# Optimization
+
+## IncOptimizer
+
+[[autodoc]] intel.neural_compressor.optimizer.IncOptimizer
```
Comment (lewtun) [inline docs/source/intel/optimization.mdx:17]: For consistency with the `onnxruntime` package, I suggest we rename this module to `intel.neural_compressor.optimization`

Comment (echarlaix) [reply docs/source/intel/optimization.mdx:17]: Yes, that's a good point, I was also thinking to rename `intel.neural_compressor.config` to `intel.neural_compressor.configuration` for the same reason. I agree that this is an important change that should be done as soon as possible (before our next release, which should be soon)

File: docs/source/intel/configuration.mdx [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+-->
+
+# Trainer
+
+## IncTrainer
+
+[[autodoc]] intel.neural_compressor.trainer_inc.IncTrainer
```
Comment (echarlaix) [inline docs/source/intel/configuration.mdx:17]: There seems to be a mismatch with `docs/source/intel/trainer.mdx`

**Review by echarlaix — APPROVED** [review]

**Comment by echarlaix** [comment]
Great addition, thanks a lot @lewtun !!

File: docs/source/intel/configuration.mdx (hunk 4 above)
Comment (lewtun) [reply docs/source/intel/configuration.mdx:17]: Yes, I wasn't sure whether to use the same hierarchy as we have in the `onnxruntime` docs. Happy to nest it though so it lines up with the source code!

File: docs/source/intel/optimization.mdx (hunk 3 above)
Comment (lewtun) [reply docs/source/intel/optimization.mdx:17]: OK great, let's tackle that in a separate PR :)

File: docs/source/intel/configuration.mdx (hunk 4 above)
Comment (echarlaix) [reply docs/source/intel/configuration.mdx]: What I meant is that `IncTrainer` is displayed in `configuration.mdx` while`IncOptimizer` is displayed in `trainer.mdx`

Comment (echarlaix) [reply docs/source/intel/configuration.mdx]: What I meant is that `IncTrainer` is displayed in `configuration.mdx` while`IncOptimizer` is displayed in `trainer.mdx`

File: .github/workflows/build_dev_documentation.yml [hunk 5]
```diff
@@ -48,7 +48,7 @@ jobs:
           cd ..
 
           cd optimum
-          pip install .[dev,onnxruntime,intel]
+          pip install -e .[dev,onnxruntime,intel]
```
Comment (michaelbenayoun) [inline .github/workflows/build_dev_documentation.yml]: Sorry, I am not very knowledgeable on the subject, but why do we need to install the packages in development mode?

Comment (lewtun) [reply .github/workflows/build_dev_documentation.yml]: Sorry, that was me trying to debug the `doc-builder` in the GitHub actions - it's removed now!

File: docs/source/intel/configuration.mdx (hunk 4 above)
Comment (lewtun) [reply docs/source/intel/configuration.mdx]: Good catch, fixed now!

File: .github/workflows/build_dev_documentation.yml [hunk 6]
```diff
@@ -88,6 +88,7 @@ jobs:
           NODE_OPTIONS: --max-old-space-size=6656
         run: |
           cd doc-build-dev && git pull
+          python -c "from optimum import intel; print(dir(intel))"
```
Comment (lewtun) [inline .github/workflows/build_dev_documentation.yml:91]: This is a temporary debug for the GH actions - will remove before merging

**Pull request merged**